pkg/tfbridge/schema.go

```diff
@@ -1022,33 +1022,64 @@ func CleanTerraformSchema(tfs map[string]*schema.Schema) map[string]*schema.Sche
 }
 
 // CoerceTerraformString coerces a string value to a Go value whose type is the type requested by the Terraform schema
-// type. Returns an error if the string can't be successfully coerced to the requested type.
-func CoerceTerraformString(schType schema.ValueType, stringValue string) (interface{}, error) {
+// type or the Pulumi SchemaInfo. We prefer the SchemaInfo overrides as it's an explicit call to action over the
+// Terraform Schema. Returns an error if the string can't be successfully coerced to the requested type.
+func CoerceTerraformString(schType schema.ValueType, ps *SchemaInfo, stringValue string) (interface{}, error) {
+	// check for the override and use that over terraform if available
+	// we do this to ensure that we are following the explicit call to action of the override
+	if ps != nil && ps.Type != "" {
+		switch ps.Type {
+		case "boolean":
+			return convertTfStringToBool(stringValue)
+		case "int":
+			return convertTfStringToInt(stringValue)
+		case "float":
+			return convertTfStringToFloat(stringValue)
+		}
+
+		return stringValue, nil
+	}
+
 	switch schType {
 	case schema.TypeInt:
-		intVal, err := strconv.ParseInt(stringValue, 0, 0)
-		if err != nil {
-			return nil, err
-		}
-		return float64(intVal), nil
+		return convertTfStringToInt(stringValue)
 	case schema.TypeBool:
-		boolVal, err := strconv.ParseBool(stringValue)
-		if err != nil {
-			return nil, err
-		}
-		return boolVal, nil
+		return convertTfStringToBool(stringValue)
 	case schema.TypeFloat:
-		floatVal, err := strconv.ParseFloat(stringValue, 64)
-		if err != nil {
-			return nil, err
-		}
-		return floatVal, nil
+		return convertTfStringToFloat(stringValue)
 	}
 
 	// Else it's just a string.
 	return stringValue, nil
 }
 
+func convertTfStringToBool(stringValue string) (interface{}, error) {
+	if stringValue == "" {
+		return false, nil // false is the default bool
```

Comment:
Is `false` really the proper default for all cases? Why would they need to introduce the string to express 3 states then? Can we actually pass `undefined`/`null`/`nil` instead?

Reply:
Yeah, I'm not sure that this is correct either.

The original behaviors used by this method were taken directly from the TF engine; we should continue this approach so we are consistent with TF's semantics.